dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -357,13 +359,42 @@ class AsyncRequestHandler : public UnaryCallback<bool>
             setDone(done_msg);
     }
 
-    bool sendPackets()
+    bool sendPackets(std::string & err_info)
     {
         for (size_t i = 0; i < read_packet_index; ++i)
         {
             auto & packet = packets[i];
-            if (!pushPacket<enable_fine_grained_shuffle, false>(request->source_index, req_info, packet, *msg_channels, log))
+            // We shouldn't throw error directly, since the caller works in a standalone thread.
+            try
+            {
+                if (!pushPacket<enable_fine_grained_shuffle, false>(
+                        request->source_index,
+                        req_info,
+                        std::make_shared<DB::TrackedMppDataPacket>(packet, current_memory_tracker),
+                        *msg_channels,
+                        log))
+                    return false;
+            }
+            catch (Exception & e)
+            {
+                err_info = e.displayText();
```

Comment:
maybe can use `getCurrentExceptionMessage`, so don't need to distinguish exception type here.

Reply:
good idea

Reply:
done